src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,536 @@
+   */
+  getSlidesAfter(slide: HTMLElement): HTMLElement[] {
+    if (this.allowsLooping()) {
+      return this.splitSlidesInHalf(slide, Direction.RIGHT);
+    } else {
+      return this.getSlides().slice(this.getSlides().indexOf(slide) + 1);
+    }
+  }
+
+  /**
+   * Transition to the next slide.
+   */
+  next(): void {
+    this.transitionSlidesBy(1);
+  }
+
+  /**
+   * Transition to the previous slide.
+   */
+  previous(): void {
+    this.transitionSlidesBy(-1);
+  }
+
+  /**
+   * Transition to the slide `value` slides away.
+   * @param value
+   */
+  transitionSlidesBy(value: number): void {
+    const nextIndex =
+        this.getSlides().indexOf(this.getCurrentTransitionTarget()) + value;
+    this.transitionToIndex(nextIndex);
+  }
+
+  /**
+   * Transition to the slide with the given index.
+   *
+   * @param index
+   * @param drivenBySync
+   */
+  transitionToIndex(index: number, drivenBySync: boolean = false): void {
+    const clampedIndex = this.getClampedIndex(index);
+    this.transitionToSlide(this.getSlideByIndex(clampedIndex), drivenBySync);
+  }
+
+  /**
+   * Register a function to be called when the given event is fired.
+   */
+  on(event: CarouselEvent, callback: (carousel: Carousel) => void) {
+    this.eventManager.on(event, callback);
+  }
+
+  /**
+   * Remove a function that is called when the given event is fired.
+   */
+  off(event: CarouselEvent, callback: (carousel: Carousel) => void) {
+    this.eventManager.off(event, callback);
+  }
+
+  /**
+   * Return the slide element at the given index in the list of slides.
+   * @param index
+   */
+  getSlideByIndex(index: number): HTMLElement {
+    return this.slides[index];
+  }
+
+  /**
+   * Return the number of slides.
+   */
+  getSlideCount(): number {
+    return this.slides.length;
+  }
+
+  /**
+   * Dispose of the carousel.
+   */
+  dispose() {
+    this.raf.dispose();
+    this.transition.dispose();
+    if (this.carouselSynchronizer !== null) {
+      this.carouselSynchronizer.dispose();
+    }
+  }
+
+  /**
+   * Clear the transition target, stop transitioning.
+   */
+  stopTransition(): void {
+    this.transitionTarget = null;
+  }
+
+  /**
+   * Clamp a given index to land within the possible slide indices.
+   * @param index
+   */
+  getClampedIndex(index: number): number {
+    const slidesLength = this.getSlides().length;
+    if (this.allowsLooping()) {
+      return mathf.wrap(index, 0, slidesLength);
+    } else {
+      return mathf.clamp(0, slidesLength - 1, index);
+    }
+  }
+
+  /**
+   * Return the index of the slide currently being transitioned to.
+   */
+  getTransitionTargetIndex(): number {
+    return this.getSlideIndex(this.getTransitionTarget().element);
+  }
+
+  /**
+   * Return the current transition target.
+   */
+  getTransitionTarget(): TransitionTarget {
+    return this.transitionTarget;
+  }
+
+  /**
+   * Synchronize this carousel with the given carousels.
+   */
+  sync(...carousels: Carousel[]): void {
```

Comment:
It seems like `this.carouselSynchronizer` is only set on the carousel that `sync()` is called on. e.g.:

```
a = new Carousel(...)
b = new Carousel(...)
a.sync(b)
a.carouselSynchronizer  // -> Object
b.carouselSynchronizer  // -> null/undefined
```

We also need some deterministic way of ensuring that any combination of `sync()` calls trigger updates to every linked carousel. I dunno if this class is easy to unit test but I'd want to ensure something like this works:

```
a.sync(b)
c.sync(d)
b.sync(c)
```

We can potentially simplify the logic of creating sync objects if we assume only 2 carousels are synced at a time (e.g. changing the method signature to `sync(carousel: Carousel)`). Then we can do something like:

* when syncing `a` and `b`, check if a sync object exists on `a` or `b` and use that if so
* if both have a `sync` object, either find some way of merging the two objects so they're linked  or take a lazy approach and just throw an error
* if no sync object exists on either, create a new sync object and add that to both `a` and `b`

Reply:
If I'm understanding the concerns correctly, these situations are already handled.

Yes only one CarouselSynchronizer instance is created for a call to sync, this class already handles bi-directional changes from the other carousels.

We can't assume only 2 carousels are synced, I currently have a use case involving 7 carousels.

I can definitely add some demos. I can probably add some unit tests, I'm not sure if we have stuff setup for testing with DOM, I can dig in.

The current big limitation is that if you have `a, b, c` and `sync(a, b)` and `sync(b, c)` a change in a will not prompt a change in c and vice-versa. This was solved by the singleton and it was agreed that allowing this case was acceptable in exchange for removing the singleton pattern.

Reply:
Right so what I'm suggesting is create a single sync object that's shared between the two carousels so that `a.carouselSynchronzer` and `b.carouselSynchronizer` both exist. And if you do `b.sync(c)` then take the synchronizer from `b` and set it to `c`.

Reply:
If we have `a, b, c, d`

`a.sync(b)` All good

`c.sync(d)` All good

`b.sync(c)` How should we handle this case? Should we handle this case?

The more we've changed this synchronizer the more cases I feel we're ignoring and it's starting to feel like a step backwards.

Reply:
See my note within carousel-synchronizer.ts. A way of conceptualizing it is if you imagine each carousel as an "island" and carousels that are synced are connected by a "bridge". If you need to pass a message to each connected island, visit each connected island and mark it as "visited". Then for every bridge on that island, check if you've already visited that island, and if not, visit that island.

If you search for "visit all linked nodes in undirected graph" you can find more examples with psuedocode.

Reply:
Sorry I misread your comment. Here's what I'm proposing:

```
a.sync(b)  // -> creates synchronizerA
a.carouselSynchronizer  // -> synchronizerA
b.caoruselSynchronzier // -> synchronizerA
```

```
c.sync(d)  // -> creates synchronizerB
c.carouselSynchronizer  // -> synchronizerB
d.caoruselSynchronzier // -> synchronizerB
```

When you do `c.sync(b)` I'd suggest one of two things: have some way of `synchronizerA` and `synchronizerB` merging its graph or simply throwing an error (which I mentioned in my original comment, "if both have a sync object, either find some way of merging the two objects so they're linked or take a lazy approach and just throw an error")

Reply:
FWIW I did mention in a previous comment that I felt it would be easier to just store a list of connected carousels within the carousel itself, in which case you wouldn't need to a way to manage the synchronizer objects.

Reply:
All good, I thought we had wanted to move away from merging sets, but I think I misunderstood the motivation of moving towards an undirected graph. Will write the merge code and then brush up on my knowledge of undirected graphs to fill in my gaps.

Reply:
CarouselSynchronizer has been removed entirely. New setup within carousel handles this case.